File: src/src_platform_jfx/fmg/jfx/utils/Timer.java
package fmg.jfx.utils;

import java.util.Date;
import java.util.function.Consumer;

import javafx.animation.Animation;
import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.util.Duration;

import fmg.common.ui.ITimer;

/** JavaFX timer. Worked in UI thread */
public class Timer implements ITimer {

    private Timeline jfxTimer;
    private long interval = 200;
    private Consumer<ITimer> callback;
    private boolean paused = true;
    /** if played: it`s start time; if paused: time offset */
    private long started;

    @Override
    public long getInterval() { return interval; }

    @Override
    public void setInterval(long delay) {
        interval = delay;
        if (jfxTimer != null)
            jfxTimer.setDelay(Duration.millis((int)interval));
    }

    private void myCallback() {
        if (callback != null)
            callback.accept(this);
    }

    @Override
    public void setCallback(Consumer<ITimer> callback) {
        this.callback = callback;
    }

    @Override
    public void start() {
        if (!paused)
            return;

        paused = false;
        started = new Date().getTime() - started; // apply of pause delta time

        if (jfxTimer == null) {
            jfxTimer = new Timeline(new KeyFrame(Duration.millis((int)interval),
                                                 ev -> myCallback()));
            jfxTimer.setCycleCount(Animation.INDEFINITE);
        }
        jfxTimer.play();
    }

    @Override
    public void pause() {
        if (paused)
            return;

        paused = true;
        started = new Date().getTime() - started; // set of pause delta time

        if (jfxTimer != null)
            jfxTimer.stop();
    }

    @Override
    public void restart() {
        pause();
        started = 0;
        start();
    }

    @Override
    public long getTime() {
        if (paused)
            return started;
        return new Date().getTime() - started;
    }

    @Override
    public void close() {
        callback = null;
        if (jfxTimer != null) {
            jfxTimer.stop();
            jfxTimer = null;
        }
    }

}

File: src/src_platform_uwp/fmg/uwp/utils/Timer.cs
using System;
using Windows.UI.Xaml;
using Fmg.Common.UI;

namespace Fmg.Uwp.Utils {

    public class Timer : ITimer {

        private DispatcherTimer _timer;
        private long _interval = 200;
        private Action<ITimer> _callback;
        private bool _paused = true;
        /// <summary> if played: it`s start time; if paused: time offset </summary>
        private long _started;

        /// <summary> in miliseconds </summary>
        public long Interval {
            get { return _interval; }
            set {
                _interval = value;
                if (_timer != null)
                    _timer.Interval = TimeSpan.FromMilliseconds(value);
            }
        }

        private void OnTick(object sender, object ev) {
            _callback?.Invoke(this);
        }

        public Action<ITimer> Callback { set { _callback = value; } }

        public void Start() {
            if (!_paused)
                return;

            _paused = false;
            _started = DateTimeOffset.Now.ToUnixTimeMilliseconds() - _started; // apply of pause delta time

            if (_timer == null) {
                _timer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(Interval) };
                _timer.Tick += OnTick;
            }
            _timer.Start();
        }

        public void Pause() {
            if (_paused)
                return;

            _paused = true;
            _started = DateTimeOffset.Now.ToUnixTimeMilliseconds() - _started; // set of pause delta time
            if (_timer != null)
                _timer.Stop();
        }

        public void Reset() {
            Pause();
            _started = 0;
        }

        public long Time {
            get {
                if (_paused)
                    return _started;
                return DateTimeOffset.Now.ToUnixTimeMilliseconds() - _started;
            }
        }

        private void Clean() {
            if (_timer == null)
                return;

            _timer.Tick -= OnTick;
            _timer.Stop();
            _timer = null;
            _callback = null;
        }

        public void Dispose() {
            Clean();
        }

    }

}
